### ISSUE TITLE
onnx 12: Inconsistent shape for ConcatLayer in function 'cv::dnn::ConcatLayerImpl::getMemoryShapes'

### ISSUE THREAD

**Issue by JulienMaille** [opencv/opencv#18695]
I'm working on a Unet model and the upsampling layers will trigger an exception when infering on an image with a different size than the training dataset: `Inconsistent shape for ConcatLayer in function 'cv::dnn::ConcatLayerImpl::getMemoryShapes'`
When exporting to onnx opset 9, inference work for any input size. The upsampling layer is defined as `Upsample`
![image](https://user-images.githubusercontent.com/182520/97590329-c6a4aa80-19fe-11eb-9fb7-59f4b9339a4f.png)

When exporting to onnx opset >9, inference work on 256² only. The upsampling layer is defined as `Resize`
![image](https://user-images.githubusercontent.com/182520/97588931-43368980-19fd-11eb-9ba6-e72b687ce29a.png)

I can provide both onnx by email if needed

**Comment by JulienMaille** [comment]
@alalek wasn't that PR #16490 supposed to handle this? https://github.com/opencv/opencv/pull/16490/files

**Comment by JulienMaille** [comment]
I confirm that editing my onnx and adding "zoom_factor_x", "zoom_factor_y" will make it work. So what we would need is to read from scales input instead

**Comment by JulienMaille** [comment]
@alalek I'd be glad to help but I don't understand why the code quoted from onnx_importer doesn't already fix the bug?

**Comment by JulienMaille** [comment]
@l-bat I do not want to break your code, but going step by step trough onnw_importer for the `Resize` layer I find strange things.
As you can see above, the note has 3 inputs: the cv::Mat, a roi, the scales (in that order)
So when you retrieve the last input here, you get the scales, not the cv::Mat
https://github.com/opencv/opencv/blob/199687a1c5b194481754d3efe14cb06a82efa805/modules/dnn/src/onnx/onnx_importer.cpp#L1683-L1690
And later when you retrieve the first input, you get the cv::Mat shape, not the scales
https://github.com/opencv/opencv/blob/199687a1c5b194481754d3efe14cb06a82efa805/modules/dnn/src/onnx/onnx_importer.cpp#L1693-L1695

Am I wrong?

**Comment by JulienMaille** [comment]
@alalek looking at the code, it seems it's more a bug than a feature request. I'm really willing to help fixing it. Can I go ahead with suggestion above?

**Comment by l-bat** [comment]
@JulienMaille you are right. We don't consider that only one of 'scales' and 'sizes' can be specified. Could you please to add `if .. else` block to your pull request?

**Comment by JulienMaille** [comment]
Hello, thank you @l-bat Just to make it clear, do you mean I should push this code to my PR?
```c++
if (node_proto.input_size() == 3)
{
    Mat scales = getBlob(node_proto, 2);
    CV_Assert(scales.total() == 4);
    layerParams.set("zoom_factor_y", scales.at<float>(2));
    layerParams.set("zoom_factor_x", scales.at<float>(3));
}
else
{
    MatShape shapes = outShapes[node_proto.input(0)];
    CV_Assert(shapes.size() == 4);
    layerParams.set("width", shapes[2]);
    layerParams.set("height", shapes[3]);
}
```

**Comment by l-bat** [comment]
Yes, I think it's easier to understand. But you need a change:
```cpp
layerParams.set("width", shapes[3]);
layerParams.set("height", shapes[2]);
```

**Comment by JulienMaille** [comment]
Sorry I edited in the mean time, with an additional question.

**Comment by l-bat** [comment]
This is strange because official ONNX docs https://github.com/onnx/onnx/blob/master/docs/Changelog.md#Resize-13 say that **Only one of 'scales' and 'sizes' can be specified.**

**Comment by JulienMaille** [comment]
Well, this onnx comes from your commit, and was sent to netron.app to look at the inputs

**Comment by JulienMaille** [comment]
> Yes, I think it's easier to understand. But you need a change:
> 
> ```c++
> layerParams.set("width", shapes[3]);
> layerParams.set("height", shapes[2]);
> ```
Ok, do you confirm original code here is correct?
https://github.com/opencv/opencv/blob/199687a1c5b194481754d3efe14cb06a82efa805/modules/dnn/src/onnx/onnx_importer.cpp#L1689-L1690

**Comment by l-bat** [comment]
If we have 3 inputs, then we need to determine the last input of the `scale` or `sizes`. I think we can determine this by the data type, integer for sizes and float for scales.

**Comment by l-bat** [comment]
According to doc [our model](https://github.com/opencv/opencv_extra/blob/master/testdata/dnn/onnx/models/dynamic_resize.onnx) is incorrect ( we are generating them for a different opset), because  **One of 'scales' and 'sizes' MUST be specified and it is an error if both are specified.** Could you please generate a new model for tests?

**Comment by JulienMaille** [comment]
> If we have 3 inputs, then we need to determine the last input of the `scale` or `sizes`. I think we can determine this by the data type, integer for sizes and float for scales.

Do you mean the input name (shown by netron) is not available in opencv onnx importer?

**Comment by l-bat** [comment]
Yes, in opencv we extract all inputs to CV::Mat

**Comment by JulienMaille** [comment]
> According to doc our model is incorrect  Could you please generate a new model for tests?

Can you provide some guidance on how to do this? I tried to edit the current one using `onnx.helper` but failed.

**Comment by alalek** [comment]
@JulienMaille Please use this script used for generation of .onnx test files: https://github.com/opencv/opencv_extra/blob/4.5.0/testdata/dnn/onnx/generate_onnx_models.py

**Comment by JulienMaille** [comment]
@alalek it's weird the python file has missing includes and wrong indenting, is it in use?
Anyway, without change, the script will pass the final size to Upsample, and generate a model with inputs [X, roi, scales, sizes] so an incorrect model according to @l-bat 

I would suggest to add a model passing scales to Upsample like this:
```python
class DynamicResizeScale(nn.Module):
    def __init__(self):
        super(DynamicResizeScale, self).__init__()

    def forward(self, x, y):
        h = y.size(2)
        w = y.size(3)
        up = nn.Upsample(scale_factor=(0.5, 0.5), mode='bilinear')
        return up(x) + y

input_0 = Variable(torch.randn(1, 3, 8, 6))
input_1 = Variable(torch.randn(1, 3, 4, 3))
model = DynamicResizeScale()
save_data_and_model_multy_inputs("dynamic_resize_scale", model, input_0, input_1, version=11)
```
![image](https://user-images.githubusercontent.com/182520/99295127-4f1bab80-2845-11eb-8d9f-e30899749686.png)

**Comment by alalek** [comment]
> we are generating them for a different opset

It is possible to generate testdata for different opset versions (see `version=`).

Also it is possible to check [ONNX version of the model](https://github.com/opencv/opencv/blob/4.5.0/modules/dnn/src/onnx/onnx_importer.cpp#L386) during loading process and adjust layers interpretation.

**Comment by JulienMaille** [comment]
With an up-to-date onnx package (1.8.0) highest opset is 12 and when exporting in 12 I still have 4 inputs with both scales and sizes
`save_data_and_model_multy_inputs("dynamic_resize", model, input_0, input_1, version=12)`

**Comment by JulienMaille** [comment]
In opset 11 resize has 3 or 4 parameters according to the manual:

X : T1
>N-D tensor

roi : T2
>1-D tensor given as [start1, ..., startN, end1, ..., endN], where N is the rank of X. The RoIs' coordinates are normalized in the coordinate system of the input image. It only takes effect when coordinate_transformation_mode is "tf_crop_and_resize"

scales : tensor(float)
>The scale array along each dimension. It takes value greater than 0. If it's less than 1, it's sampling down, otherwise, it's upsampling. The number of elements of 'scales' should be the same as the rank of input 'X'. Only one of 'scales' and 'sizes' can be specified. If 'size' is needed, the user can use an empty string as the name of 'scales' in this operator's input list.

sizes (optional) : tensor(int64)
>The size of the output tensor. The number of elements of 'sizes' should be the same as the rank of input 'X'. Only one of 'scales' and 'sizes' can be specified.

**Comment by JulienMaille** [comment]
Ok so I export both models with different opset, this is what I got
```
when giving size & opset  9 -> Upsample [X, scales]
when giving size & opset 10 -> Resize   [X, scales]
when giving size & opset 11 -> Resize   [X, roi, scales, sizes]

when giving scales & opset  9 -> Upsample [X, scales]
when giving scales & opset 10 -> Resize   [X, scales]
when giving scales & opset 11 -> Resize   [X, roi, scales]
```
